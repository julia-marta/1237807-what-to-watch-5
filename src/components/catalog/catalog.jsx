import React from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";
import {changeGenre, showMoreCards, resetCards} from "../../store/actions";
import MovieList from "../movie-list/movie-list";
import GenresList from "../genres-list/genres-list";
import ShowMoreButton from "../show-more-button/show-more-button";
import withActiveCard from "../../hocs/with-active-card/with-active-card";
import {getGenre, getCardsCount, getFilteredFilms, getGenres} from "../../store/selectors";
import movieCardProp from "../../prop-types/movie-card.prop";

const MovieListWrapped = withActiveCard(MovieList);

const Catalog = (props) => {
  const {filteredFilms, genresList, activeGenre, changeGenreAction, cardsCount, showMoreCardsAction, resetCardsAction} = props;
  const renderedFilms = filteredFilms.slice(0, cardsCount);
  const renderedFilmsCount = renderedFilms.length;

  return (
    <section className="catalog">
      <h2 className="catalog__title visually-hidden">Catalog</h2>

      <GenresList genres={genresList} activeGenre={activeGenre} onGenreClick={changeGenreAction} resetCards={resetCardsAction} />

      <MovieListWrapped films={renderedFilms} />

      {filteredFilms.length > renderedFilmsCount ?
        <ShowMoreButton onShowMoreButtonClick={showMoreCardsAction} filmsToShowCount={filteredFilms.slice(renderedFilmsCount).length}/>
        : ``}
    </section>
  );
};

Catalog.propTypes = {
  filteredFilms: PropTypes.arrayOf(movieCardProp).isRequired,
  genresList: PropTypes.arrayOf(PropTypes.string).isRequired,
  activeGenre: PropTypes.string.isRequired,
  changeGenreAction: PropTypes.func.isRequired,
  cardsCount: PropTypes.number.isRequired,
  showMoreCardsAction: PropTypes.func.isRequired,
  resetCardsAction: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  filteredFilms: getFilteredFilms(state),
  genresList: getGenres(state),
  activeGenre: getGenre(state),
  cardsCount: getCardsCount(state),
});

const mapDispatchToProps = (dispatch) => ({
  changeGenreAction(genre) {
    dispatch(changeGenre(genre));
  },
  showMoreCardsAction(filmsToShowCount) {
    dispatch(showMoreCards(filmsToShowCount));
  },
  resetCardsAction() {
    dispatch(resetCards());
  }
});

export {Catalog};
export default connect(mapStateToProps, mapDispatchToProps)(Catalog);
