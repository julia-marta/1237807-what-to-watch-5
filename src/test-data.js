import React from 'react';
import PropTypes from "prop-types";
export const noop = () => {};

export const promoFilm = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  posterImage: `the-grand-budapest-hotel-poster.jpg`,
  backgroundImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#000000`,
  genre: `Drama`,
  released: 2014,
  isFavorite: false,
};

export const filmCard = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  genre: `Drama`,
  previewImage: `the-grand-budapest-hotel.jpg`,
  previewVideoLink: `https://trailers.apple.com/movies/universal/krampus/krampus-tlr1_h480p.mov`,
};

export const film = {
  id: 1,
  name: `The Grand Budapest Hotel`,
  posterImage: `the-grand-budapest-hotel-poster.jpg`,
  backgroundImage: `bg-the-grand-budapest-hotel.jpg`,
  backgroundColor: `#000000`,
  genre: `Drama`,
  released: 2014,
  description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.
  Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
  rating: 8.9,
  scoresCount: 240,
  director: `Wes Andreson`,
  starring: [`Bill Murray`, `Edward Norton`, `Jude Law`, `Willem Dafoe`],
  runTime: 99,
  isFavorite: true,
};

export const reviews = [
  {
    id: 1,
    user: {
      id: 4,
      name: `Kate Muir`,
    },
    rating: 8.9,
    comment: `Discerning travellers and Wes Anderson fans will luxuriate in the glorious Mittel-European kitsch of one of the director's funniest and most exquisitely designed movies in years.`,
    date: `2019-05-08T14:13:56.569Z`,
  },
  {
    id: 2,
    user: {
      id: 5,
      name: `Matthew Lickona`,
    },
    rating: 7.2,
    comment: `The mannered, madcap proceedings are often delightful, occasionally silly, and here and there, gruesome and/or heartbreaking.`,
    date: `2019-05-18T14:13:56.569Z`,
  }
];

export const genres = [`All genres`, `Drama`, `Music`, `Biography`, `Survival`, `Thriller`, `Crime`];

export const defaultState = {
  DATA: {
    films: [filmCard, filmCard, filmCard, filmCard, filmCard, filmCard, filmCard],
    promo: promoFilm,
    currentFilm: film,
    currentReviews: reviews,
  },
  FILTER: {
    genre: `Drama`,
    cardsCount: 8,
  },
  USER: {
    status: `NOT_AUTHORIZED`,
    name: `userName`,
    avatar: `avatar.jpg`,
    reviewStatus: `NOT_SAVING`,
  },
};

export const MockComponent = (props) => {
  const {children} = props;

  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

export const SimpleMockComponent = () => <div />;
