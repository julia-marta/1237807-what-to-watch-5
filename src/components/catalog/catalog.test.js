import React from "react";
import renderer from "react-test-renderer";
import {BrowserRouter} from 'react-router-dom';
import {Catalog} from "./catalog";
import {noop, genres, defaultState} from "../../test-data";

const mockFilteredFilms = defaultState.DATA.films;

describe(`should Catalog render correctly`, () => {
  it(`with ShowMoreButton`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <Catalog filteredFilms={mockFilteredFilms} genresList={genres}
              activeGenre={genres[0]} changeGenreAction={noop} cardsCount={4}
              showMoreCardsAction={noop} resetCardsAction={noop}/>
          </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
  it(`without ShowMoreButton`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <Catalog filteredFilms={mockFilteredFilms} genresList={genres}
              activeGenre={genres[0]} changeGenreAction={noop} cardsCount={mockFilteredFilms.length}
              showMoreCardsAction={noop} resetCardsAction={noop}/>
          </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
  it(`without filtered films for active genre`, () => {
    const tree = renderer
      .create(
          <BrowserRouter>
            <Catalog filteredFilms={[]} genresList={genres}
              activeGenre={genres[2]} changeGenreAction={noop} cardsCount={4}
              showMoreCardsAction={noop} resetCardsAction={noop}/>
          </BrowserRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
