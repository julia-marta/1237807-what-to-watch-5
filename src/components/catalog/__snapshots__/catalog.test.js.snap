// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should Catalog render correctly with ShowMoreButton 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Drama
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Music
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Biography
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Survival
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Thriller
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Crime
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
  </div>
  <div
    className="catalog__more"
  >
    <button
      className="catalog__button"
      onClick={[Function]}
      type="button"
    >
      Show more
    </button>
  </div>
</section>
`;

exports[`should Catalog render correctly without ShowMoreButton 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Drama
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Music
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Biography
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Survival
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Thriller
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Crime
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  >
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
    <article
      className="small-movie-card catalog__movies-card"
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
    >
      <div
        className="small-movie-card__image"
        onClick={[Function]}
      >
        <img
          alt="The Grand Budapest Hotel"
          height="175"
          src="the-grand-budapest-hotel.jpg"
          width="280"
        />
      </div>
      <h3
        className="small-movie-card__title"
      >
        <a
          className="small-movie-card__link"
          href="/films/1"
          onClick={[Function]}
        >
          The Grand Budapest Hotel
        </a>
      </h3>
    </article>
  </div>
  
</section>
`;

exports[`should Catalog render correctly without filtered films for active genre 1`] = `
<section
  className="catalog"
>
  <h2
    className="catalog__title visually-hidden"
  >
    Catalog
  </h2>
  <ul
    className="catalog__genres-list"
  >
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        All genres
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Drama
      </a>
    </li>
    <li
      className="catalog__genres-item catalog__genres-item--active"
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Music
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Biography
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Survival
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Thriller
      </a>
    </li>
    <li
      className="catalog__genres-item "
    >
      <a
        className="catalog__genres-link"
        onClick={[Function]}
      >
        Crime
      </a>
    </li>
  </ul>
  <div
    className="catalog__movies-list"
  />
  
</section>
`;
